src/Functions/sparseGrams.cpp

```diff
@@ -27,83 +27,150 @@ extern const int NUMBER_OF_ARGUMENTS_DOESNT_MATCH;
+    std::vector<std::pair<size_t, size_t>> result;
+    size_t iter_result = 0;
+
+    struct PositionAndHash
+    {
+        size_t position;
+        size_t left_ngram_position;
+        size_t hash;
+    };
+
+    class NGramSymbolIterator
     {
-        if constexpr (is_utf8)
+    public:
+        NGramSymbolIterator() = default;
+
+        NGramSymbolIterator(Pos data_, Pos end_, size_t n_)
+            : data(data_), end(end_), n(n_)
+        {
+        }
+
+        void increment()
+        {
+            right_iterator = getNextPosition(right_iterator);
+
+            if (++num_increments >= n)
+                left_iterator = getNextPosition(left_iterator);
+        }
+
+        bool isEnd() const
+        {
+            return data + right_iterator >= end;
+        }
+
+        std::pair<size_t, size_t> getNGramPositions() const
+        {
+            return {left_iterator, right_iterator};
+        }
+
+        size_t getNextPosition(size_t iterator) const
         {
-            Poco::UTF8Encoding encoder{};
-            size_t byte_offset = 0;
-            while (pos + byte_offset < end)
+            if constexpr (is_utf8)
+                return iterator + UTF8::seqLength(data[iterator]);
+            else
+                return iterator + 1;
+        }
+
+    private:
+
+        Pos data;
+        Pos end;
+        size_t n;
+        size_t right_iterator = 0;
+        size_t left_iterator = 0;
+        size_t num_increments = 0;
+    };
+
+    /// The convex hull contains the maximum values ​​of the suffixes that start from the current right iterator.
+    /// For example, if we have n-gram hashes like [1,5,2,4,1,3] and current right position is 4 (the last one)
+    /// than our convex hull will consists of elements:
+    /// [{position:1, hash:5}, {position:3, hash:4}, {position:4,hash:1}]
+    /// Assuming that hashes are uniformly distributed, the expected size of convex_hull is N^{1/3},
+    /// where N is the length of the string.
+    /// Proof: https://math.stackexchange.com/questions/3469295/expected-number-of-vertices-in-a-convex-hull
+    std::vector<PositionAndHash> convex_hull;
+    NGramSymbolIterator symbol_iterator;
+
+    /// Get the next batch of answers. Returns false if there can be no more answers.
+    bool consume()
+    {
+        if (symbol_iterator.isEnd())
+            return false;
+
+        auto [ngram_left_position, right_position] = symbol_iterator.getNGramPositions();
+        size_t next_right_position = symbol_iterator.getNextPosition(right_position);
+        size_t right_border_ngram_hash = hasher(pos + ngram_left_position, next_right_position - ngram_left_position);
+
+        while (!convex_hull.empty() && convex_hull.back().hash < right_border_ngram_hash)
+        {
+            size_t possible_left_position = convex_hull.back().left_ngram_position;
+            size_t length = next_right_position - possible_left_position;
+            if (length > max_ngram_length)
```

Comment:
This and the next length check are wrong for UTF8 characters, check out the last two tests, it shouldn't return the empty list